dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/BackwardsReachingDefinitionsAnalysis.java

```diff
@@ -0,0 +1,384 @@
+ * Especially, it does not collect last writers for all registers.
+ * <p>
+ * This analysis is control-flow-sensitive;
+ * that is, {@link Label} splits the information among the jumps to it
+ * and {@link CondJump} merges the information.
+ * <p>
+ * This analysis supports loops;
+ * that is, backward jumps cause re-evaluation of the loop body until convergence.
+ * This results in a squared worst-case time complexity in terms of events being processed.
+ */
+public class BackwardsReachingDefinitionsAnalysis implements ReachingDefinitionsAnalysis {
+
+    private final Map<RegReader, ReaderInfo> readerMap = new HashMap<>();
+    private final Map<RegWriter, Readers> writerMap = new HashMap<>();
+
+    @Override
+    public Writers getWriters(RegReader reader) {
+        Preconditions.checkNotNull(reader, "reader is null");
+        final ReaderInfo result = readerMap.get(reader);
+        Preconditions.checkArgument(result != null, "reader %s has not been analyzed.", reader);
+        return result;
+    }
+
+    @Override
+    public Writers getFinalWriters() {
+        final ReaderInfo result = readerMap.get(null);
+        Preconditions.checkState(result != null, "final state has not been analyzed.");
+        return result;
+    }
+
+    /**
+     * Lists all potential users of the result of an event.
+     * @param writer Event of interest to fetch information about.
+     * @throws IllegalArgumentException If {@code writer} was not inside the program when it was analyzed.
+     */
+    public Readers getReaders(RegWriter writer) {
+        Preconditions.checkNotNull(writer, "writer is null");
+        final Readers result = writerMap.get(writer);
+        Preconditions.checkArgument(result != null, "writer %s has not been analyzed.", writer);
+        return result;
+    }
+
+    /**
+     * Lists all potential users of uninitialized registers.
+     */
+    public Readers getInitialReaders() {
+        final Readers result = writerMap.get(null);
+        Preconditions.checkState(result != null, "initial state has not been analyzed.");
+        return result;
+    }
+
+    /**
+     * Analyzes an entire function definition.
+     * @param function Defined function to be analyzed.  Its parameters are marked as initialized.
+     */
+    public static BackwardsReachingDefinitionsAnalysis forFunction(Function function) {
+        final var analysis = new BackwardsReachingDefinitionsAnalysis();
+        final Set<Register> finalRegisters = new HashSet<>();
+        analysis.initialize(function, finalRegisters);
+        analysis.run(function, finalRegisters);
+        analysis.postProcess();
+        return analysis;
+    }
+
+    /**
+     * Analyzes an entire set of threads.
+     * <p>
+     * Optionally queries {@link ExecutionAnalysis} for pairs of writers appearing together in an execution.
+     * @param program Contains a set of threads to be analyzed.  Additionally-defined functions are ignored.
```

Comment:
It sounds like the analysis is intended to be run only after threads has been created. If so, maybe we should add some check somewhere.

EDIT: Now I am confused. The comment says that functions are ignored, but below you have one case where you iterate over both threads and functions.